tests/cupy_tests/linalg_tests/test_solve.py

```diff
@@ -148,6 +148,65 @@ def test_invalid_shape(self):
         self.check_shape((4, 3, 2, 1), rcond=0.1)
 
 
+@unittest.skipUnless(
+    cuda.cusolver_enabled, 'Only cusolver in CUDA 8.0 is supported')
+@testing.gpu
+class TestLstsq(unittest.TestCase):
+
+    @testing.for_float_dtypes(no_float16=True)
+    @condition.retry(10)
+    def check_x(self, a_shape, b_shape, rcond, dtype):
+        a_cpu = numpy.random.randint(0, 10, size=a_shape).astype(dtype)
+        b_cpu = numpy.random.randint(0, 10, size=b_shape).astype(dtype)
+        a_gpu = cupy.asarray(a_cpu)
+        b_gpu = cupy.asarray(b_cpu)
+        a_gpu_copy = a_gpu.copy()
+        b_gpu_copy = b_gpu.copy()
+        result_cpu, resids_cpu, rank_cpu, s_cpu = numpy.linalg.lstsq(a_cpu,
+                                                                     b_cpu,
+                                                                     rcond=rcond)  # noqa E501
+        result_gpu, resids_gpu, rank_gpu, s_gpu = cupy.linalg.lstsq(a_gpu,
+                                                                    b_gpu,
+                                                                    rcond=rcond)  # noqa E501
+
+        self.assertEqual(result_cpu.dtype, result_gpu.dtype)
+        cupy.testing.assert_allclose(result_cpu, result_gpu, atol=1e-3)
+        cupy.testing.assert_allclose(resids_cpu, resids_gpu, atol=1e-3)
+        cupy.testing.assert_allclose(rank_cpu, rank_gpu, atol=1e-3)
+        cupy.testing.assert_allclose(s_cpu, s_gpu, atol=1e-3)
+        cupy.testing.assert_array_equal(a_gpu_copy, a_gpu)
+        cupy.testing.assert_array_equal(b_gpu_copy, b_gpu)
```

Comment:
I check if ||ax-b||^2 is close with ```cupy.testing.assert_allclose(resids_cpu, resids_gpu, atol=1e-3)```

I'm not sure about ```atol=1e-3```, I picked this because it was the same value for tests of ```cupy.linalg.pinv``` which also uses svd.

Reply:
Does not this implementation ensure that the result (first element of return value) of `cupy.lstsq` is equal to that of `numpy.lstsq`?

Reply:
Why `@condition.retry(10)` at L157 is needed? One of the result value is often distant from the expected value?

Reply:
> Why @condition.retry(10) at L157 is needed? One of the result value is often distant from the expected value?

Long response https://github.com/cupy/cupy/pull/2165#issuecomment-490572406